src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -1049,17 +1056,150 @@ public static bool TryDecode(ImmutableArray<byte> bytes)
+
+            UnicodeCategory cat = CharUnicodeInfo.GetUnicodeCategory(ch);
+            return IsLetterChar(cat)
+                || IsDecimalDigitChar(cat)
+                || IsConnectingChar(cat)
+                || IsCombiningChar(cat)
+                || IsFormattingChar(cat);
+        }
+
+
+        private static bool IsLetterChar(UnicodeCategory cat)
+        {
+            // letter-character:
+            //   A Unicode character of classes Lu, Ll, Lt, Lm, Lo, or Nl 
+            //   A Unicode-escape-sequence representing a character of classes Lu, Ll, Lt, Lm, Lo, or Nl
+
+            switch (cat)
+            {
+                case UnicodeCategory.UppercaseLetter:
+                case UnicodeCategory.LowercaseLetter:
+                case UnicodeCategory.TitlecaseLetter:
+                case UnicodeCategory.ModifierLetter:
+                case UnicodeCategory.OtherLetter:
+                case UnicodeCategory.LetterNumber:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsCombiningChar(UnicodeCategory cat)
+        {
+            // combining-character:
+            //   A Unicode character of classes Mn or Mc 
+            //   A Unicode-escape-sequence representing a character of classes Mn or Mc
+
+            switch (cat)
+            {
+                case UnicodeCategory.NonSpacingMark:
+                case UnicodeCategory.SpacingCombiningMark:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsDecimalDigitChar(UnicodeCategory cat)
+        {
+            // decimal-digit-character:
+            //   A Unicode character of the class Nd 
+            //   A unicode-escape-sequence representing a character of the class Nd
+
+            return cat == UnicodeCategory.DecimalDigitNumber;
+        }
+
+        private static bool IsConnectingChar(UnicodeCategory cat)
+        {
+            // connecting-character:  
+            //   A Unicode character of the class Pc
+            //   A unicode-escape-sequence representing a character of the class Pc
+
+            return cat == UnicodeCategory.ConnectorPunctuation;
+        }
+
+        /// <summary>
+        /// Returns true if the Unicode character is a formatting character (Unicode class Cf).
+        /// </summary>
+        /// <param name="cat">The Unicode character.</param>
+        private static bool IsFormattingChar(UnicodeCategory cat)
+        {
+            // formatting-character:  
+            //   A Unicode character of the class Cf
+            //   A unicode-escape-sequence representing a character of the class Cf
+
+            return cat == UnicodeCategory.Format;
+        }
+
+        /// <summary>
         /// Given an input string changes it to be acceptable as a part of a type name.
-        /// For now we will simply replace '.' with '_'as the most common case.
+        /// See Github #1428 for discussion
```

Comment:
We don't place github references in doc comments. This link will be meaningless in most contexts; this is never formatted as markdown.